corehq/apps/domain/calculations.py

```diff
@@ -164,10 +164,13 @@ def j2me_forms_in_last_bool(domain, days):
 
 
 def _sms_helper(domain, direction=None, days=None):
+    assert direction in (INCOMING, OUTGOING, None)
     query = SMSES().domain(domain).size(0)
 
-    if direction:
-        query = query.direction(direction)
+    if direction is INCOMING:
+        query = query.incoming_messages()
+    elif direction is OUTGOING:
```

Comment:
This is a string identity comparison while the assertion above is doing equality comparisons, which means the assertion may pass while both of these evaluate to false unexpectedly. This could happen if someone passes in a `direction` value that is equal to but has a different identity (memory address) than `INCOMING` or `OUTGOING` constants. Python does try to cache and reuse some strings, but it does not guarantee that all strings that compare equal will have the same identity. It would be safer to use `==` rather than `is` here:
```py
    if direction == INCOMING:
        query = query.incoming_messages()
    elif direction == OUTGOING:
```

See this [SO answer](https://stackoverflow.com/a/21203303/10840) for more details. Incidentally I am unable to construct an example that fails on Python 3, and I believe that is because the particular values used in these constants are aggressively interned because they consist of only characters allowed in Python identifiers. However, that does not mean this is not fragile.

Reply:
Thanks for the thoughtful review! I'll make this change. Not sure why I originally chose to go with `is`, but I agree `==` is the better choice for comparison with strings (though your answer taught me more about why, so thanks!)

Reply:
I also just had to prove to myself that a few lines above, the `in` operator internally compares with `==` and not `is`, so the `assert` isn't vulnerable to the same thing. 😌